### PR TITLE
Guidance hint solution that resolves #2007

### LINKED ISSUE #2007: Support guidance form for hints
From https://opendatakit.github.io/xforms-spec/#languages:

> Even within a single language, it is helpful to have multiple ‘forms’ of the same string. For example, a verbose phrasing used as the caption when answering a question, but a short, terse phrasing when that question is shown in the form summary. This can be done using the form attribute.

> Supported for <hint> content only. It is a description of the question that can be used to provide further guidance to enumerators. It is not meant to be shown in the client UI by default, but could be shown in a special view mode or on printouts.

I propose adding a setting at the bottom of General settings > form management > form filling labeled `Show guidance for questions` with possible values `Yes` and `No`. It should be `No` by default. If it is set to `Yes`, guidance text will be shown below the hint in blue italics.

Behavior discussion at https://forum.opendatakit.org/t/collect-collapsible-training-hints-guidance/12247

### PR DESCRIPTION
**Pull request by jd-alexander**
Closes #2007 

#### What has been done to verify that this works as intended?

To verify that this functionality works I manually tested it by creating a form that has guidance attached to the hint of its question. Once that was done, I loaded the form into the Collect app, enabled guidance hints to be shown on questions within the Form Management Settings and opened the form's question which displayed the hint. 

To further verify this functionality I wrote an Espresso test that enabled the guidance setting and I created a test instance of `FormEntryActivity` with the guidance hint form. I then fetched the guidance hint from the respective `FormEntryPrompt` and compared it with the text of the TextView on the screen that's supposed to perform the rendering. 

#### Why is this the best possible solution? Were any other approaches considered?
This is the best possible solution due to the nature of how the JavaRosa SDK allows text strings with form attributes to be fetched. 

#### Are there any risks to merging this code? If so, what are they?
The only risks associated with merging this code would be the changes that were made to the `HelpTextView`'s layout constraints that were associated with the rendering of the `AnswerView.` This was being done entirely in code so a xml layout file was used to encapsulate the additional layout that was necessary for the guidance text view. Nonetheless, I made several checks to ensure that these changes didn't break the functionality of the questions and the preexisting UI tests along with the UI test that was written for this feature verifies that this won't cause an impact.

#### Do we need any specific form for testing your changes? If so, please attach one.
To test this functionality you need a form that has a [guidance hint](https://opendatakit.github.io/xforms-spec/#languages) eg. 
[guidance_hint_form.xml.txt](https://github.com/opendatakit/collect/files/1888351/guidance_hint_form.xml.txt)

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
https://github.com/opendatakit/docs/issues/635

### CODE REVIEW

File: collect_app/src/main/res/layout/help_text_layout.xml
Comment (lognaturel) [inline collect_app/src/main/res/layout/help_text_layout.xml]: This makes all hint text that wraps centered which doesn't look right. See All Widgets form to see what I mean!

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: There shouldn't be any line breaks here.

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: Please remove extra line breaks! I saw a few other places where there were extra ones so would be good to double check the diff.

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: Our style doesn't call for a line break here.

File: collect_app/src/main/res/values/strings.xml
Comment (lognaturel) [inline collect_app/src/main/res/values/strings.xml]: Should be in sentence case ("Show guidance for questions").

File: collect_app/src/main/res/values/strings.xml
Comment (lognaturel) [inline collect_app/src/main/res/values/strings.xml]: always shown (all lowercase)

File: collect_app/src/main/res/values/strings.xml
Comment (lognaturel) [inline collect_app/src/main/res/values/strings.xml]: collapsed (all lowercase)

**Review by lognaturel — COMMENTED** [review]
Looking great to me! I've marked a few quick things inline as I saw them but haven't done a full review yet. 

Let's wait a day or two for feedback about the UI: https://forum.opendatakit.org/t/collect-collapsible-training-hints-guidance/12247/3

I find the (i) icon very difficult to hit even on a Galaxy Tab A tablet. Is it possible only the actual path is tappable and not the transparent parts? Could we perhaps expand the tappable area?

I wasn't sure initially that I liked the icon but I haven't come up with anything better so it has grown on me.

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java
Comment (jd-alexander) [reply collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: No prob :) will definitely keep an eye out for them!

**Comment by jd-alexander** [comment]
Thanks for the review! 
Icon/UI - Yes I am not too sure about it either so I just put it in place until we could all decide on something together. I have an idea to spruce up the UI a bit so I am gonna make those changes and share them here. 

Tappable area 
- > I find the (i) icon very difficult to hit even on a Galaxy Tab A tablet. Is it possible only the actual path is tappable and not the transparent parts? Could we perhaps expand the tappable area?

Yes, we could make clicks only take place on the actual path. Currently I just have an `OnClickListener` on the `TextView` because I am using it's `CompoundDrawable` mechanism  so no matter the length of the hint the icon will always be properly aligned at the end of the text. Started working on a layout with a `TextView` and `ImageView` but I just thought this approach would yield the same results.

**Comment by lognaturel** [comment]
> Currently I just have an OnClickListener on the TextView

Of course, I hadn't noticed that! That explains why I was having a tough time when I tried to tap the (i). Eventually it would be good to have both text and icon be tappable but let's wait for some preliminary UI feedback. 😊

**Comment by jd-alexander** [comment]
Okay :) I made just the drawable clickable for now. Gonna do some thorough testing of it to see how it behaves.

File: gradle/wrapper/gradle-wrapper.properties
Comment (yanokwa) [inline gradle/wrapper/gradle-wrapper.properties]: I think this change should be reverted because we don't need the sources and documentation that `all` comes with. Why download things we don't need 😂

**Comment by yanokwa** [comment]
I think this works really nicely, and we have tests! Hard to do complain about that...

Some changes I'd like to see before this goes to testing.
* The `i` icon should probably be black. Gray makes it look disabled.
* The icon isn't a huge touch surface, so maybe we can make the whole hint text clickable, to make life easier for users? This is similar to how you can select the entire line for a radio button in selects ones.

I'm not a huge fan of the `i` after the text (although @lognaturel prefers it!). I had similar pushback when we added the red star for required questions (https://github.com/opendatakit/collect/pull/1828#pullrequestreview-94182756) and I was able to convince folks to move the red star to the beginning. I'm not necessarily arguing for a change, but I'm curious how you decided to put it at the end, @jd-alexander.

File: gradle/wrapper/gradle-wrapper.properties
Comment (jd-alexander) [reply gradle/wrapper/gradle-wrapper.properties]: haha no prob. Never knew that's what the `all` meant. Thanks for pointing that out.

**Comment by jd-alexander** [comment]
Good looking out @yanokwa . Honestly, I was used to putting asterisks/icons about hints at the end of the text from my old job so that's the reason I placed it there. I did some UX research on this topic and I found some [detailed explanations from the folks over at UX Stack Exchange](https://ux.stackexchange.com/questions/56023/when-a-button-contains-text-and-an-icon-which-should-come-first) that justifies it being placed before the text so I am gonna opt for that approach because it makes more sense and it actually lessens most of the issues I had to deal with such as the icon trailing the text and causing inconsistencies with other hint guidance's that might be shorter. I am gonna work on this change and a subtle effect to make this more appealing and share it here :smile:

**Comment by yanokwa** [comment]
@jd-alexander If you can get those changes sent in soon and fix the conflicts, I think we'd be ready to test and merge!

**Comment by jd-alexander** [comment]
@yanokwa Yes I am gonna finish this up! Was just a bit pre occupied with the SMS stuff but yes I want to finalize this so I will spend some time doing this tomorrow!

**Comment by jd-alexander** [comment]
@yanokwa I was thinking of adding a background to the guidance hint when it's in collapse mode so when it gets expanded it has a bit more body and outline. I have this working example in my repo of how it would look when it's open. What do you think?
I included this in the pushes I sent so you can take a look if necessary. 

Collapsed Hint Guidance

**Comment by jd-alexander** [comment]
Hey @yanokwa I submitted the changes. Not sure what's causing this build to fail though. Probably triggering a rebuild might cause it to pass because there are no style errors.

File: collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/preferences/PreferenceKeys.java]: Indentation here should match that of other values.

File: collect_app/src/main/java/org/odk/collect/android/preferences/GuidanceHint.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/preferences/GuidanceHint.java]: Remove extra blank line.

File: collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java]: Remove extra blank line

File: collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java]: Remove extra blank line

File: collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java]: Remove extra blank line

File: collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java:132]: Remove extra blank line

File: collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/AnimateUtils.java]: Remove extra blank line

File: collect_app/src/main/java/org/odk/collect/android/utilities/UiUtils.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/UiUtils.java]: Remove "This method"

File: collect_app/src/main/java/org/odk/collect/android/utilities/UiUtils.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/UiUtils.java]: Remove "This method"

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: Remove extra blank line

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: Remove extra blank line

File: collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java]: Remove extra blank line

**Review by lognaturel — COMMENTED** [review]
Few small things and this will be ready for QA! 

- The guidance hint shouldn't disappear on rotate
- The hint font size is already small so I would like guidance hints to have the same font size

I don't find the box around the guidance hint necessary but I don't feel very strongly about it. @yanokwa?

**Comment by yanokwa** [comment]
I don't love the box.

**Comment by jd-alexander** [comment]
Okay no prob! Will remove the box!

**Comment by lognaturel** [comment]
Awesome! Once you've done the changes and the merge conflicts are resolved (sorry!), you can add the `needs-testing` tag with `@opendatakit-bot label needs-testing` and it will go through the QA process.

**Comment by jd-alexander** [comment]
@opendatakit-bot label needs-testing

**Comment by jd-alexander** [comment]
- Removed box outline
- Added collapsed state to saved state
- Integrated control with dark theme
- Corrected style errors

Ready for testing :fireworks:

**Comment by mmarciniak90** [comment]
@jd-alexander I noticed a problem in case:
1. User installs the application (`Show guidance for questions` is set as a `No` by default)
2. User starts `Guidance Form Sample` form
3. User goes to `General Settings`
4. User changes `Show guidance for questions` to `Yes - collapsed`
5. User backs to form
6. User sees error


Additionally, `i` icon is cut off on devices with >5 inches screen
screenshot_2018-05-08-16-09-47

**Comment by lognaturel** [comment]
Good catch, @mmarciniak90! @jd-alexander, hopefully these are relatively quick fixes. Would be great to get this in a beta by Thursday!

**Comment by jd-alexander** [comment]
Thanks for catching these @mmarciniak90  @lognaturel They should be! Working on them now.

**Comment by jd-alexander** [comment]
@mmarciniak90 resolved the issues! Ready for another test to verify.

**Comment by mmarciniak90** [comment]
@jd-alexander mentioned problems are resolved!
I checked more question types and I think that repeating guidance for `Full date and time` question type is not expected.
screenshot_2018-05-18-12-52-01
You can use form: [hints.xml.txt](https://github.com/opendatakit/collect/files/1987313/hints.xml.txt)

**Comment by jd-alexander** [comment]
@mmarciniak90 Resolved and ready to go!

**Comment by mmarciniak90** [comment]
**Tested with success!**

@jd-alexander great work! Everything looks great!

Verified on Android 4.1, 4.2, 4.4., 5.1 6.0, 7.0, 8.1

Verified cases:
* clear settings
* many types of question
* full date and time question
* many questions on one page
* dark theme

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

**Review by lognaturel — APPROVED** [review]

**Pull request merged**
